WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/Downloadable Files/FIle List/ProductDownloadListViewModelTests.swift

```diff
@@ -0,0 +1,163 @@
+
+        // Act
+        let viewModel = ProductDownloadListViewModel(product: model)
+
+        // Assert
+        XCTAssertEqual(viewModel.count(), 0)
+        XCTAssertEqual(viewModel.downloadLimit, -1)
+        XCTAssertEqual(viewModel.downloadExpiry, -1)
+    }
+
+    // MARK: - `handleDownloadableFilesChange`
+
+    func testHandlingADuplicateDownloadableFilesUpdatesWithError() {
+
+    }
+
+    func testHandlingAValidDownloadableFilesUpdatesWithSuccess() {
+
+    }
+
+    func testHandlingTheOriginalDownloadableFilesIsAlwaysValidAndUnique() {
+
+    }
+
+    func testHandlingAValidDownloadableFilesAddsWithSuccess() {
+
+    }
+
+    func testHandlingAValidExistingDownloadableFileRemovesWithSuccess() {
+
+    }
+
+    // MARK: - `handleDownloadLimitChange`
+
+    func testHandlingAValidDownloadLimitUpdatesWithSuccess() {
+        // Arrange
+        let product = MockProduct().product(downloadable: true)
+        let model = EditableProductModel(product: product)
+
+        // Act
+        let viewModel = ProductDownloadListViewModel(product: model)
+        viewModel.handleDownloadLimitChange(5)
+
+        // Assert
+        XCTAssertEqual(viewModel.downloadLimit, 5)
+    }
+
+    // MARK: - `handleDownloadExpiryChange`
+
+    func testHandlingAValidDownloadExpiryUpdatesWithSuccess() {
+        // Arrange
+        let product = MockProduct().product(downloadable: true)
+        let model = EditableProductModel(product: product)
+
+        // Act
+        let viewModel = ProductDownloadListViewModel(product: model)
+        viewModel.handleDownloadExpiryChange(5)
+
+        // Assert
+        XCTAssertEqual(viewModel.downloadExpiry, 5)
+    }
+
+    // MARK: - `hasUnsavedChanges`
+
+    func testViewModelHasUnsavedChangesAfterUpdatingDownloadFileLimit() {
+        // Arrange
+        let product = MockProduct().product(downloadable: true)
+        let model = EditableProductModel(product: product)
+
+        // Act
+        let viewModel = ProductDownloadListViewModel(product: model)
+        viewModel.handleDownloadLimitChange(5)
+
+        // Assert
+        XCTAssertTrue(viewModel.hasUnsavedChanges())
+
+    }
+
+    func testViewModelHasUnsavedChangesAfterUpdatingDownloadFileExpiry() {
+        // Arrange
+        let product = MockProduct().product(downloadable: true)
+        let model = EditableProductModel(product: product)
+
+        // Act
+        let viewModel = ProductDownloadListViewModel(product: model)
+        viewModel.handleDownloadExpiryChange(5)
+
+        // Assert
+        XCTAssertTrue(viewModel.hasUnsavedChanges())
+    }
+
+    func testViewModelHasUnsavedChangesAfterUpdatingDownloadFilesOrder() {
+        // Arrange
+        let product = MockProduct().product(downloadable: true)
+        let model = EditableProductModel(product: product)
+
+        // Act
+        let viewModel = ProductDownloadListViewModel(product: model)
+        guard let firstFile = viewModel.remove(at: 0) else {
+            XCTFail()
```

Comment:
nit: I suggest to add a comment on why it's failing inside the `XCTFail` case

Reply:
yeap. added